libuv -> uvw overhaul

jagerman opened the pull request:
Refactors a bunch of code around the event loop/logic thread to make everything use `uvw` (which is much nicer to work with, but is also needed to plug the plainquic code into lokinet).

Based on top of the android branch #1537 because it touched a bunch of the same areas in the DNS code.

Draft because I have another commit half done to merge the Logic code into the EventLoop code.

jagerman committed:
Add more junk & clangd symlinks to .gitignore

Ignore any root directory starting with build so that you can have many
build directories (build, build2, build-win32, ...) that get ignored.

Add .cache which clangd creates

Ignore compile_commands.json (and generate it by default) which helps
clangd, but needs to be symlinked to a compile_commands.json in a build
dir so isn't suitable for inclusion in the repo.

jagerman committed:
Bump required oxenmq version

Current lokinet dev relies on something from oxenmq dev branch which is
in the not-yet-released next version; without this lokinet tries using
the system 1.2.3 install, which then fails the build.

jagerman committed:
Fix signed/unsigned char comparison bug

These functions would not working properly if given a high bit char
value.

jagerman committed:
Shut up noisy catch tests

jagerman committed:
Tweak clang-tidy rules

Turn off modernize-use-trailing-return because it's *really* annoying
(and clangd picks it up when using it for editing).

Turn off modernize-use-nodiscard as well because it's already annoying.

Add bugprone-* as well because it includes a lot of useful checks.

jagerman committed:
Disallow evil compilation

-fpermissive allows nasty things that are (apparently) needed on
Windows, so only enable it there.

vla are evil, so don't just warn but error on it.

remove duplicate -Wno-unknown-pragmas

jagerman committed:
Replace libuv with uvw & related refactoring

- removes all the llarp_ev_* functions, replacing with methods/classes/functions in the llarp
  namespace.
- banish ev/ev.h to the void
- Passes various things by const lvalue ref, especially shared_ptr's that don't need to be copied
  (to avoid an atomic refcount increment/decrement).
- Add a llarp::UDPHandle abstract class for UDP handling
- Removes the UDP tick handler; code that needs tick can just do a separate handler on the event
  loop outside the UDP socket.
- Adds an "OwnedBuffer" which owns its own memory but is implicitly convertible to a llarp_buffer_t.
  This is mostly needed to take over ownership of buffers from uvw without copying them as,
  currently, uvw does its own allocation (pending some open upstream issues/PRs).
- Logic:
  - add `make_caller`/`call_forever`/`call_every` utility functions to abstract Call wrapping and
    dependent timed tasks.
  - Add inLogicThread() so that code can tell its inside the logic thread (typically for
    debugging assertions).
  - get rid of janky integer returns and dealing with cancellations on call_later: the other methods
    added here and the event loop code remove the need for them.
- Event loop:
  - redo everything with uvw instead of libuv
  - rename EventLoopWakeup::Wakeup to EventLoopWakeup::Trigger to better reflect what it does.
  - add EventLoopRepeater for repeated events, and replace the code that reschedules itself every
    time it is called with a repeater.
  - Split up `EventLoop::run()` into a non-virtual base method and abstract `run_loop()` methods;
    the base method does a couple extra setup/teardown things that don't need to be in the derived class.
  - udp_listen is replaced with ev->udp(...) which returns a new UDPHandle object rather that
    needing gross C-style-but-not-actually-C-compatible structs.
  - Remove unused register_poll_fd_(un)readable
  - Use shared_ptr for EventLoopWakeup rather than returning a raw pointer; uvw lets us not have to
    worry about having the event loop class maintain ownership of it.
  - Add factory EventLoop::create() function to create a default (uvw-based) event loop (previously
    this was one of the llarp_ev_blahblah unnamespaced functions).
  - ev_libuv: this is mostly rewritten; all of the glue code/structs, in particular, are gone as
    they are no longer needed with uvw.
- DNS:
  - Rename DnsHandler to DnsInterceptor to better describe what it does (this is the code that
    intercepts all DNS to the tun IP range for Android).
- endpoint:
  - remove unused "isolated network" code
  - remove distinct (but actually always the same) variables for router/endpoint logic objects
- llarp_buffer_t
  - make constructors type-safe against being called with points to non-size-1 values
- tun packet reading:
  - read all available packets off the device/file descriptor; previously we were reading one packet
    at a time then returning to the event loop to poll again.
  - ReadNextPacket() now returns a 0-size packet if the read would block (so that we can implement
    the previous point).
  - ReadNextPacket() now throws on I/O error
- Miscellaneous code cleanups/simplifications

jagerman committed:
C++17 simplification

jagerman committed:
Merge Logic functions into EventLoop

loop->call(...) is similar to the old logic->Call(...), but is smart
about the current thread: if called from within the event loop it simply
runs the argument directly, otherwise it queues it.

Similarly most of the other event loop calls are also now thread-aware:
for example, `call_later(...)` can queue the job directly when called if
in the event loop rather than having to double-queue through the even
loop (once to call, then inside the call to initiate the time).

jagerman committed:
Make format

jagerman committed:
Switch uvw to a fork with libuv backwards compat support

jagerman committed:
Remove unused variable

jagerman committed:
Fix linux compilation & narrowing conversion warning

jagerman committed:
Remove unused junk (also fix win32 compilation fail)

jagerman committed:
Rename EventLoop::udp to EventLoop::make_udp

Makes it consistent with the make_waker, etc.

Also removes some debugging.

tewinget commented on llarp/link/server.cpp:
```diff
@@ -125,18 +126,19 @@ namespace llarp
-      static_cast<ILinkLayer*>(udp->user)->RecvFrom(from, std::move(pkt));
-    };
-    m_udp.tick = &ILinkLayer::udp_tick;
+    LogError("omg wtf");
```
nice meme

tewinget commented on llarp/link/server.cpp:
```diff
@@ -335,35 +339,29 @@ namespace llarp
 
   bool
-  ILinkLayer::Start(std::shared_ptr<Logic> l)
+  ILinkLayer::Start(const EventLoop_ptr& loop)
```
should have m_Loop already set here, so argument no longer necessary.

tewinget commented on llarp/router/outbound_message_handler.cpp:
```diff
@@ -184,8 +184,8 @@ namespace llarp
     {
-      auto f = std::bind(callback, status);
-      LogicCall(_logic, [self = this, f]() { self->m_Killer.TryAccess(f); });
+      auto f = [f = std::move(callback), status] { f(status); };
```
So uhh...un-f this I guess?

jagerman committed:
Simplify code of many f's

We don't need to go through the contention killer here, that was mainly
for (long ago) debugging reasons.

tewinget commented on llarp/router/rc_lookup_handler.cpp:
```diff
@@ -157,7 +157,7 @@ namespace llarp
     {
       LogDebug("Adding or updating RC for ", RouterID(rc.pubkey), " to nodedb and dht.");
       const RouterContact copy{rc};
```
looks like this copy is no longer needed then?

jagerman committed:
Remove unused arguments (and ants)

jagerman committed:
Android & hive fixes

jagerman committed:
Remove HasNextPacket() (and fix win32)

We don't use it anymore (instead we just ReadNextPacket until it returns
an empty result).

jagerman committed:
Export UDP file descriptor (for android)

jagerman committed:
Make llarp_buffer_t constructible from an rvalue reference

jagerman committed:
Uncurse SockAddr's sockaddr* operator

The `const sockaddr*` conversion was returning a pointer to the
sockaddr_in6, but with its family set to AF_INET, so the receiving code
would have treated interpreting the result as a sockaddr_in pointer --
but it isn't, and trying to get the IPv4 from the mis-reinterpreted
sockaddr_in6 fails; this then prevented uvw UDP binding (which uses the
sockaddr* operator) from working.

This changes it to always leave the family value at the proper value and
using the ipv6_is_mapped_ipv4 check to determine when we have (and
should return) a pointer to the sockaddr_in rather than the
sockaddr_in6.

Also fixes toString() to compare the full prefix instead of just bytes
10 and 11.

jagerman committed:
Remove unused ShouldFlushNow method

jagerman committed:
EventLoop: add public wakeup() method, and call it from call()

call(), when invoked from the logic thread, wasn't triggering a wakeup
which stalled some traffic (such as client-to-snode packets).

Fix it by triggering a wakeup on `call()`, and expose it because this is
a useful thing to be able to do.

jagerman committed:
Tick event loop on data queuing

jagerman committed:
Modernize [[maybe_unused]] attribute

jagerman committed:
Add wakeup call into transit_hop UpstreamHandler
